drivers/rtt_rtc/rtt_rtc.c

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright (C) 2020 Benjamin Valentin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @ingroup     drivers_periph_rtc
+ * @{
+ *
+ * @file
+ * @brief       Basic RTC implementation based on a RTT
+ *
+ * @note        Unlike a real RTC, this emulated version is not guaranteed to keep
+ *              time across reboots or deep sleep.
+ *              If your hardware provides the means to implement a real RTC, always
+ *              prefer that over this emulated version!
+ *
+ * @author      Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ *
+ * @}
+ */
+
+#include <stdlib.h>
+#include <string.h>
+
+#include "periph/rtc.h"
+#include "periph/rtt.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#define RTT_SECOND      (RTT_FREQUENCY)
+
+#define _RTT(n)         ((n) & RTT_MAX_VALUE)
+
+#define RTT_SECOND_MAX  (RTT_MAX_VALUE/RTT_FREQUENCY)
+
+#define TICKS(x)        (    (x) * RTT_SECOND)
+#define SECONDS(x)      (_RTT(x) / RTT_SECOND)
+
+#define _NOINIT         __attribute__((section(".noinit")))
+
+static uint32_t alarm_time;     /*< The RTC timestamp of the (user) RTC alarm */
+static uint32_t rtc_now _NOINIT;/*< The RTC timestamp when the last RTT alarm triggered */
```

Comment:
With `_NOINIT`, `rtc_now` has an arbitrary value until it is set for the first time. Is this the desired behavior?

I guess `rtc_now` should survive reboots or deep sleep if possible, right? If the CPU implementation has some RAM which survives reboots or deep sleeps, it should be used here. For example, on ESPs `__attribute__((section(".rtc.bss")))` would have to be used here. On other CPUs it might be `__attribute__((section(".backup.bss")))`.

I'm wondering whether it would be better to define a special macro in `drivers/include/periph/rtt.h`, or even more globally, that a CPU could override, for example
```
#ifndef PERSIST_RAM_ATTR /* any better name is welcome */
#define PERSIST_RAM_ATTR __attribute__((section(".noinit")))
#endif
````
```suggestion
static uint32_t rtc_now PERSIST_RAM_ATTR; /*< The RTC timestamp when the last RTT alarm triggered */
```

Reply:
> With `_NOINIT`, `rtc_now` has an arbitrary value after power on until it is set for the first time. Is this the desired behavior?

Well I'm afraid we don't have a generic way to detect power-on reset yet.

> If the CPU implementation has some RAM which survives reboots or deep sleeps, it should be used here.

I agree!

> I'm wondering whether it would be better to define a special macro in `drivers/include/periph/rtt.h`, or even more globally, that a CPU could override, for example
> 
> ```
> #ifndef PERSIST_RAM_ATTR /* any better name is welcome */
> #define PERSIST_RAM_ATTR __attribute__((section(".noinit")))
> #endif
> ```

Absolutely. And this is a property of the MCU, it's not RTT specific, so it should be in `periph_cpu.h`/`periph_cpu_common.h`.
`tests/periph_backup_ram` would use it too.

Reply:
> > With `_NOINIT`, `rtc_now` has an arbitrary value after power on until it is set for the first time. Is this the desired behavior?
> 
> Well I'm afraid we don't have a generic way to detect power-on reset yet.

I know. For ESPs it is realized in that way, that `.rtc.bss` is cleared only on power-up reset. That is, if the value in `.rtc.bss` is zero, it is from power-up reset. I supposed other platforms do it in same way.